File: 백준/Gold/17143. 낚시왕/낚시왕.java
import java.awt.List;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Stack;

public class Main {
	static int r, c, m;
	static int result = 0;

	static class go {

		int x, y, velo, dir, size;

		public go(int x, int y, int velo, int dir, int size) {
			super();
			this.x = x;
			this.y = y;
			this.velo = velo;
			this.dir = dir;
			this.size = size;
		}

		public go(int x, int y) {
			super();
			this.x = x;
			this.y = y;
		}

		public go(int x, int y, int velo, int dir) {
			super();
			this.x = x;
			this.y = y;
			this.velo = velo;
			this.dir = dir;
		}

	}

	static ArrayList<go> shark;

	public static void main(String[] args) throws IOException {

		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

		StringTokenizer s = new StringTokenizer(in.readLine(), " ");

		r = Integer.parseInt(s.nextToken());
		c = Integer.parseInt(s.nextToken());
		m = Integer.parseInt(s.nextToken());
		shark = new ArrayList<>();
		for (int i = 0; i < m; i++) {

			s = new StringTokenizer(in.readLine(), " ");
			int x, y, vel, dir, size;
			x = Integer.parseInt(s.nextToken());
			x--;
			y = Integer.parseInt(s.nextToken());
			y--;
			vel = Integer.parseInt(s.nextToken());
			dir = Integer.parseInt(s.nextToken());
			size = Integer.parseInt(s.nextToken());
			shark.add(new go(x, y, vel, dir, size));

		}

		game();
		System.out.println(result);

	}

	private static void game() {
		// TODO Auto-generated method stub

		for (int i = 0; i < c; i++) {

			// 낚시왕과 가장 가까운 상어를 한마리 잡음
			// System.out.println(i);
			eat(i);

			// 상어를 이동시킴

			move();
			// System.out.println();
			// 상어가 중복된 곳에 존재를 할시 가장 큰 상어를 제외하고 제거

		}

	}

	private static void move() {
		// TODO Auto-generated method stub

		int graph[][] = new int[r][c];
		int k1 = 0;
		ArrayList<go> temp = new ArrayList<>();
		int dx[] = { 0, -1, 1, 0, 0 };
		int dy[] = { 0, 0, 0, 1, -1 };
		for (go t : shark) {

			// 방향에 따라 움직여야 함

			// d가 1인 경우는 위, 2인 경우는 아래, 3인 경우는 오른쪽, 4인 경우는 왼쪽
			// System.out.println(t.x + " " + t.y + " " + t.dir);
			if (t.dir == 1) {
				int zx = t.x;
				int zy = t.y;
				int z1 = t.velo;
				

				while (z1 > 0) {

					int tem = dx[t.dir] * z1;

					if (tem + zx < 0) {
						int aa=Math.abs(zx-0);
						z1-=aa;
						zx=0;
						t.dir=2;

					} else if (tem + zx >= r) {
						int aa=Math.abs(r-1-zx);
						z1-=aa;
						zx=r-1;
						t.dir=1;
						
					}
					else
					{
						zx=zx+tem;
						z1=0;
					}


				}
				t.x = zx;
				t.y = zy;

			} else if (t.dir == 2) {
				int zx = t.x;
				int zy = t.y;
				int z1 = t.velo;
				while (z1 > 0) {

					int tem = dx[t.dir] * z1;

					if (tem + zx < 0) {
						int aa=Math.abs(zx-0);
						z1=z1-aa;
						zx=0;
						t.dir=2;

					} else if (tem + zx >= r) {
						int aa=Math.abs(r-1-zx);
						z1=z1-aa;
						zx=r-1;
						t.dir=1;
					}
					else
					{
						zx=zx+tem;
						z1=0;
					}


				}
				t.x = zx;
				t.y = zy;

			} else if (t.dir == 3) {
				// 오른쪽
				int zx = t.x;
				int zy = t.y;
				int z1 = t.velo;
				while (z1 > 0) {

					int tem = dy[t.dir] * z1;

					if (tem + zy < 0) {
						int aa=Math.abs(zy-0);
						z1-=aa;
						zy=0;
						t.dir=3;

					} else if (tem + zy >= c) {
						int aa=Math.abs(c-1-zy);
						z1-=aa;
						zy=c-1;
						t.dir=4;
					}
					else
					{
						zy=zy+tem;
						z1=0;
					}

				}
				t.x = zx;
				t.y = zy;

			} else if (t.dir == 4) {
				// 왼쪽
				int zx = t.x;
				int zy = t.y;
				int z1 = t.velo;

				while (z1 > 0) {

					int tem = dy[t.dir] * z1;

					if (tem + zy < 0) {
						int aa=Math.abs(zy-0);
						z1-=aa;
						zy=0;
						t.dir=3;

					} else if (tem + zy >= c) {
						int aa=Math.abs(c-1-zy);
						z1-=aa;
						zy=c-1;
						t.dir=4;
					}
					else
					{
						zy=zy+tem;
						z1=0;
					}


				}
				
				t.x = zx;
				t.y = zy;

			}
			// System.out.println(t.x+" "+t.y);
			graph[t.x][t.y] += 1;
			if (graph[t.x][t.y] == 2) {
				temp.add(new go(t.x, t.y));
			}

		}
//		for(int i=0;i<r;i++)
//			System.out.println(Arrays.toString(graph[i]));
//		System.out.println();
//		

		for (go g : temp) {
			// System.out.println("아하");
			// 중복된 상어들의 좌표만 존재를 함
			int x1 = g.x;
			int y1 = g.y;
			int size = Integer.MIN_VALUE;
			ArrayList<Integer> op = new ArrayList<>();
			for (int u = 0; u < shark.size(); u++) {
				go g1 = shark.get(u);
				if (g1.x == x1 && g1.y == y1) {
					op.add(u);
					size = Math.max(g1.size, size);
				}
			}

			// 이제 op에 들은거 삭제를 해야함
			for (int u = shark.size() - 1; u >= 0; u--) {
				go g1 = shark.get(u);
				if (g1.x == x1 && g1.y == y1) {
					if (g1.size != size) {
						shark.remove(u);
					}
				}
			}
		}

	}

	private static void eat(int now) {
		// TODO Auto-generated method stub

		int get = -1;
		int index = Integer.MAX_VALUE;
		for (int i = 0; i < c; i++) {

			for (int j = 0; j < shark.size(); j++) {
				go l = shark.get(j);
				if (l.y == now) {
					if (l.x < index) {
						index = l.x;
						get = j;
					}
				}

			}

		}

		if (get != -1) {
			go t = shark.get(get);
			result += t.size;
			// System.out.println(t.size);
			shark.remove(t);
		}

	}

}

File: 백준/Gold/17143. 낚시왕/낚시왕.py
import sys
from collections import deque

input = sys.stdin.readline

r, c, m = map(int, input().split())
graph = [[0] * (c + 1) for _ in range(r + 1)]
shark = [[[] for _ in range(c + 1)] for _ in range(r + 1)]

for _ in range(m):
    x, y, speed, dir, size = map(int, input().split())
    shark[x][y].append([speed, dir, size])
    # 1위 2아래 3오른 4왼
ans = 0


def death(z):
    global ans
    for i in range(1, r + 1):
        if len(shark[i][z]) == 1:
            speed, dir, size = shark[i][z][0]
            shark[i][z] = []
            ans += size
            break


def go(x, y, speed, dir):
    # 1위 2아래 3오른 4왼
    dx = [-1, -1, 1, 0, 0]
    dy = [-1, 0, 0, 1, -1]
    if dir == 3 or dir == 4:
        nex = (speed) % ((c - 1) * 2)
        # 이제 남은 만큼 이동 해야함
        zx = x
        zy = y
        while 1:
            tx = zx
            ty = zy
            zx += dx[dir] * nex
            zy += dy[dir] * nex
            if zy <= 0 or zy >= c + 1:
                if dir == 3:
                    dir = 4
                else:
                    dir = 3
                if zy <= 0:
                    nex -= ty
                    zy = 2
                else:
                    nex -= (c - ty + 1)
                    zy = c - 1
                if nex <= 0:
                    break
            else:
                break
    else:
        nex = (speed) % ((r - 1) * 2)
        zx = x
        zy = y
        while 1:
            tx = zx
            ty = zy
            zx += dx[dir] * nex
            zy += dy[dir] * nex
            if zx <= 0 or zx >= r + 1:
                if dir == 1:
                    dir = 2
                else:
                    dir = 1
                if zx <= 0:
                    nex -= tx
                    zx = 2
                else:
                    nex -= (r - tx + 1)
                    zx = r - 1
                if nex <= 0:
                    break
            else:
                break
    return [zx, zy, dir]


temp = [[[] for _ in range(c + 1)] for _ in range(r + 1)]


def move():
    # 이제 하나하나 이동을 하자
    for i in range(1, r + 1):
        for j in range(1, c + 1):
            if len(shark[i][j]) == 1:
                speed, dir, size = shark[i][j][0]
                x, y, dd = go(i, j, speed, dir)
                # 이제 이동 하게 됨
                temp[x][y].append([speed, dd, size])
                shark[i][j] = []

    for i in range(1, r + 1):
        for j in range(1, c + 1):
            # 상어 잡아 먹기
            if len(temp[i][j]) > 1:
                ee = temp[i][j]
                ee = sorted(ee, key=lambda x: -x[2])
                shark[i][j] = [ee[0]]
                temp[i][j] = []
            elif len(temp[i][j]) == 1:
                ee = temp[i][j]
                shark[i][j] = [ee[0]]
                temp[i][j] = []


for z in range(1, c + 1):
    # 낚시왕 이동
    # 이제 잡을 거임
    death(z)
    # 상어 이동
    move()
    # for i in shark:
    #     print(i)
    # print()
print(ans)